File: src/main/ClientHandler/ClientHandler.java
package main.ClientHandler;

import main.HTTP.Request.RequestDTO;
import main.HTTP.Response.ResponseDTO;
import main.HTTP.Builder.HttpResponseBuilder;
import main.HTTP.Parser.HTTPRequestParser;

import java.io.IOException;
import java.net.Socket;
import java.util.Map;
import java.util.function.BiConsumer;

import static main.Server.Server.router;

/**
 * The ClientHandler class is responsible for handling individual client connections.
 */
public class ClientHandler implements Runnable {

    private final Socket clientSocket;

    /**
     * Constructs a new ClientHandler with the given client socket.
     *
     * @param socket The client socket to handle.
     */
    public ClientHandler(Socket socket) {
        this.clientSocket = socket;
    }

    /**
     * Handles the client connection by processing the request and sending a response.
     */
    @Override
    public void run() {
        try {
            RequestDTO request = HTTPRequestParser.parse(clientSocket.getInputStream());
            ResponseDTO response = new ResponseDTO();

            BiConsumer<RequestDTO, ResponseDTO> handler = router.getHandler(request.getMethod(), request.getPath());

            if (handler != null) {
                handler.accept(request, response);
            } else {
                response.setStatusCode(404);
                response.setBody("Not Found you idiot");
                response.setHeaders(Map.of("Content-Type", "text/html"));
            }

            HttpResponseBuilder.build(response, clientSocket.getOutputStream());
        } catch (IOException e) {
            System.out.println("Error handling client: " + e.getMessage());
        } finally {
            try {
                clientSocket.close();
            } catch (IOException e) {
                System.out.println("Failed to close the socket: " + e.getMessage());
            }
        }
    }
}

File: src/main/java/main/ClientHandler/ClientHandler.kt
package main.java.main.ClientHandler

import main.Router.Router
import main.WebSocket.WebSocketClient
import main.java.main.HTTP.Parser.HTTPParser
import java.net.Socket
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardOpenOption

class ClientHandler(private val client: Socket): Thread() {

    companion object {
        lateinit var webSocketClient: WebSocketClient

        fun getStoredClient(clientId: String): WebSocketClient? {
            val clientsData = Files.readAllLines(Paths.get("clients.txt"))
            return clientsData
                .find { it.split(",")[0] == clientId }
                ?.let { WebSocketClient() }
        }
    }

    override fun run() {
        currentThread().name = "ClientHandler-${client.port}"
        webSocketClient = WebSocketClient()
        webSocketClient.connect()
        writeConnection(webSocketClient)
        try {
            val request = HTTPParser().parse(client.getInputStream())
            /*val response = HTTPBuilder().build(ResponseDTO(200,
                "All is well",
                mapOf(
                    Pair("Content-Type", "text/html"),
                    Pair("Upgrade", "websocket"),
                    Pair("Connection", "Upgrade")
                ),
                "<html><body><h1>No Route Found!</h1><script>$js</script></body></html>"),
                client.getOutputStream())*/
            Router().route(request, client)
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            client.close()
        }
    }

    fun writeConnection(client: WebSocketClient) {
        val clientData = mapOf(
            "id" to client,
            "timestamp" to System.currentTimeMillis().toString()
        )

        Files.write(
            Paths.get("clients.txt"),
            "${clientData["id"]},${clientData["timestamp"]}\n".toByteArray(),
            StandardOpenOption.APPEND,
            StandardOpenOption.CREATE
        )
    }
}